[Feature] Add basic stories

rosamaggi opened the pull request:
- Add input fields stories like: Radiobutton, Text, Numeric, Dropdown list,...
- Add Responsive tabs, Badge and teaser
- Active the onChange listener to most of them

domq committed:
[rewrite] Part of `@epfl/epfl-sti-react-library` into TypeScript

- Use [this named slot trick](https://medium.com/swlh/bring-vue-named-slots-to-react-87684188f18e) to give access to top-level menu, drawer, aside (left-hand) menu, avatar and breadcrumbs
- Provide reasonable defaults for (some of) them

domq committed:
[refactor] Extract `src/utils/children.ts`

domq committed:
[feature] Unbundle Avatar widget from `Base.User` locus in `Base`

- Rename `Base.UserAvatar` to `Base.User`, so that one may decide to put up a login button up there instead
- Develop a proper `<Avatar>` component, with a `<Avatar.Image>` named slot that (pretends to) accept either children or a `peopleSciper` prop

rosamaggi committed:
[Refactor] Move files to subdirectory

rosamaggi committed:
[Feature] Add text

rosamaggi committed:
[Fix] Modify label and value couple

rosamaggi committed:
[Typo]

rosamaggi committed:
[Feature] Add textarea

rosamaggi committed:
[Fix] Resize textarea

rosamaggi committed:
[Feature] Add teaser

rosamaggi committed:
[Feature] Add styles

rosamaggi committed:
[Feature] Add tabs

rosamaggi committed:
[Feature] Add Bedge

rosamaggi committed:
[Feature] Add Dropdown

rosamaggi committed:
[Feature] Add Tabs and responsive tabs

- Tabs makes tabs visible for mobile phones and cards visible for bigger screens
- ResponsiveTabs shows always tabs for every screen size

rosamaggi committed:
[Feature - WIP] Add Autocomplete input

rosamaggi committed:
[Fix] Multiple checkboxes

rosamaggi committed:
[Fix] Add onclick

rosamaggi committed:
[Feature -WIP] Add dropDown stories and onChange listener

rosamaggi committed:
[Refactor] clear code

rosamaggi committed:
[Feature] Add table of content

rosamaggi committed:
[Feature] Add onChange event and new stylesheet for autocomplete

rosamaggi committed:
[Feature] Add onChange event for checkbox

rosamaggi committed:
[Feature] Add onChange event for dropdown list

rosamaggi committed:
[Feature] Add onChange event for Numeric input

rosamaggi committed:
[Feature] Add onChange event for Radiobutton

rosamaggi committed:
[Feature] Add onChange event for TextArea input

rosamaggi committed:
[Feature] Add onChange event for Text input

rosamaggi committed:
[Feature] Add new stylesheet for Tabs

rosamaggi committed:
[Feature] Add new stylesheet for ResponsiveTabs

rosamaggi committed:
[Feature] Add new stylesheet for badge

rosamaggi committed:
[Feature] Modify stylesheet for button

rosamaggi committed:
[Refactor] change meta signature and delete unused import

rosamaggi committed:
[Refactor] Delete example files

rosamaggi committed:
[Refactor] delete custom styles sheet

rosamaggi committed:
[Feature] Add inLine property to checkboxes

rosamaggi committed:
[Fix] Add type Item for input parameter

rosamaggi committed:
[Fix] Add storybook style sheet to fix the .tag style from element-epfl

rosamaggi committed:
[Fix] Restrict tag class scope

This fixes a CSS clash with the 'element.css', as seen for instance on the screenshot in the
pull request https://github.com/epfl-si/elements/pull/702

rosamaggi committed:
[Fix] Add empty line at the end of files

domq committed:
[Install] react-bootstrap

```
yarn add react-bootstrap
```

domq reviewed (changes requested):
This is a preliminary review; I'll be  more throrough once you have made the nested-element changes, which are a big chunk of work.

rosamaggi committed:
[Fix] Add onClick event on the div of the Badge

rosamaggi commented on src/stories/molecules/Badge.tsx:
```diff
@@ -0,0 +1,55 @@
+  icon?: string;
+  textColor?: ColorValue;
+  fontWeight: 'normal' | 'bold';
+  onClickIcon?: () => void;
```
Done: the onClick event is now on the div of the Badge

rosamaggi committed:
[Fix] Add readonly props for Autocompete

rosamaggi committed:
[Fix] Rename "sizeDropDown" to "size"

rosamaggi commented on src/stories/molecules/inputFields/Dropdown.tsx:
```diff
@@ -0,0 +1,79 @@
+}
+
+interface DropdownProps {
+  sizeDropDown?: 'small' | 'medium' | 'large';
```
Done

rosamaggi committed:
[Fix] Add description to components

rosamaggi commented on src/stories/molecules/Badge.tsx:
```diff
@@ -0,0 +1,55 @@
+}
+
+/**
+ * Primary UI component for user interaction
```
Done

rosamaggi committed:
[Feature] Add sub-components

- Tab
- Title
- Content

rosamaggi commented on src/stories/molecules/Tabs.tsx:
```diff
@@ -0,0 +1,41 @@
+
+type Item = {
+  title: string;
+  tabContent?: React.ReactNode;
```
Done

rosamaggi commented on src/stories/molecules/ResponsiveTabs.stories.tsx:
```diff
@@ -0,0 +1,43 @@
+export const Default: Story = {
+  args: {
+    items: [
+      { title: "1", content: <div>
```
Done

rosamaggi commented on src/stories/molecules/Badge.tsx:
```diff
@@ -0,0 +1,55 @@
+interface BadgeProps {
+  title?: string;
+  image?: string;
+  border?: 'none' | 'solid' | 'dashed' | 'dotted';
```
![image](https://github.com/epfl-si/epfl-elements-react/assets/95354173/702771b6-ffe5-4b40-8a2d-ef0ff372d72a)

rosamaggi replied:
Done

rosamaggi committed:
[Install] Add csstype

`npm install csstype`

rosamaggi committed:
[Feature] Add csstype as props of components

rosamaggi committed:
[Fix] Styles

rosamaggi commented on yarn.lock:
```diff
@@ -4328,7 +4328,7 @@ __metadata:
   linkType: hard
 
-"@vitejs/plugin-react@npm:4.2.1":
+"@vitejs/plugin-react@npm:^4.2.1":
```
Done together in call

rosamaggi commented:
All changes have been done. Could you verify again please?

domq committed:
[Install] Bootstrap and SASS

```
yarn add bootstrap
yarn add -D sass
```

domq committed:
[Refactor] Use Bootstrap-(and SASS)-style media breakpoints

- Use Bootstrap's `@include media-breakpoint-up(XX)` and `@include media-breakpoint-down(XX)` macro as per their [documentation](https://getbootstrap.com/docs/4.6/layout/overview/#responsive-breakpoints)
- Rename CSS files to `.scss` to benefit from such compile-time macro-expansion (currently there is only one such file)
- Provide foolproof `@import '/scss/bootstrap.scss';` form to initialize all of Bootstrap SASS things in the proper order

domq reviewed (changes requested):
Another batch of work... Sorry need to run for lunch; will do thorough review yet later.

domq commented on src/stories/molecules/Button.stories.tsx:
```diff
@@ -1,34 +1,31 @@
-    </Button>
-  }
+      </svg>,
+      <span key="label" style={{ marginLeft: '5px' }}>Primary</span>,
```
I suppose the button text could be made a story parameter here?

domq commented on src/stories/molecules/Tabs.tsx:
```diff
@@ -0,0 +1,70 @@
+
+  function c_title(children: React.ReactNode) {
+    const title = (Children.toArray(children || []).find(
+      (child : React.ReactElement) => child.type === Tabs.Tab.Title)) as React.ReactElement<TabTitleProps>;
```
The type after `as` should have `[]` at the end, just like on line 14.

rosamaggi replied:
Should we have `[] `even if I used `.find` that return only one element? In case there are more than one title, I get the first one.

domq commented on src/stories/molecules/Tabs.tsx:
```diff
@@ -0,0 +1,70 @@
+
+  function c_content(children: React.ReactNode) {
+    const content = (Children.toArray(children || []).find(
+      (child : React.ReactElement) => child.type === Tabs.Tab.Content)) as React.ReactElement<TabContentProps>;
```
Same.

domq commented on src/stories/molecules/inputFields/Autocomplete.tsx:
```diff
@@ -0,0 +1,142 @@
+
+/**
+ * An autocomplete list.
+ * If single selection: the selected element is shown directly in the input field. The "itemValue" could be the
```
Please skip a line to isolate §.

I don't understand the sentences on lines 23–24. Could you please rewrite it without a conditional (“could”)?

domq commented on src/stories/molecules/inputFields/Autocomplete.tsx:
```diff
@@ -0,0 +1,142 @@
+   placeholder = '',
+   isReadonly = false,
+  onChange
+  }: AutocompleteProps & FormControlProps) => {
```
You should be able to remove the type either here, or on line 28, and let TypeScript figure it out.

domq commented on src/stories/molecules/inputFields/Autocomplete.tsx:
```diff
@@ -0,0 +1,142 @@
+  onChange
+  }: AutocompleteProps & FormControlProps) => {
+  const [filteredSuggestions, setFilteredSuggestions] = useState<Item[]>([]);
+  const [inputValue, setInputValue] = useState(itemValue ? itemValue.label : '');
```
Consider using `itemValue?.label` as the state initializer.

domq commented on src/stories/molecules/inputFields/Autocomplete.tsx:
```diff
@@ -0,0 +1,142 @@
+  }: AutocompleteProps & FormControlProps) => {
+  const [filteredSuggestions, setFilteredSuggestions] = useState<Item[]>([]);
+  const [inputValue, setInputValue] = useState(itemValue ? itemValue.label : '');
+  const [selectedSuggestions, setSelectedSuggestions] = useState<Item[]>(selected ? selected : []);
```
Likewise, I don't think you need a ternary operator here as the props initializer already provides a default value.

domq reviewed (changes requested):
Aaaaand here comes part deux.

rosamaggi committed:
[Fix] Modify comments

rosamaggi committed:
[Fix] Delete redundant type declaration

rosamaggi committed:
[Fix] Add onChange check before doing function logic

rosamaggi committed:
[Fix] Add global variables for background and shadow

rosamaggi committed:
[Refactor] Split style to be more readable

rosamaggi committed:
[Refactor] Delete dead code

rosamaggi committed:
[Fix] Replace hrefTeaser and hrefRightIcon with onClick  on all the teaser object

rosamaggi committed:
[Fix] Replace left icon image and delete dead code about hrefs

rosamaggi committed:
[Refactor] Add new line

rosamaggi committed:
[Fix] add correct type declaration for stories

rosamaggi committed:
[Fix] Change style name to not to have "storybook" in the name and move switch block inside the else (if the button size is not icon)

rosamaggi committed:
[Fix] Add bedge prop

rosamaggi committed:
[Fix] Add react import

domq committed:
[Improve] `Button` story: use explicit `render()` function

This lets us tweak the text in the button (which is not a prop).

⚠ There is a rather stringent set of poorly-documented requirements:
- a [suitable `argTypes`](https://storybook.js.org/docs/api/arg-types) structure must be set up in the `meta` object
- In `meta.argTypes`, the `defaultValue` is not just [deprecated](https://storybook.js.org/docs/api/arg-types#defaultvalue); it doesn't actually work at all — One must pass the default values through `args` in the story instead
- [the `render()` function must take an argument](https://github.com/storybookjs/storybook/issues/23343#issuecomment-1627351756)
- (As stated in a comment in the code) The set of props passed to the component-under-test must be exactly stated (otherwise the generated code will understandingly be confused)

domq committed:
[Refactor] Simplify `Button` story with args composition

As documented in the [args composition pattern](https://storybook.js.org/docs/writing-stories/args#args-composition)

rosamaggi committed:
[Refactor] Use Arg composition to clean code

https://storybook.js.org/docs/writing-stories/args#args-object

rosamaggi committed:
[Refactor] Move Card component from frontend to epfl-element-react

domq committed:
[fix] Whole `<Badge>` is clickable ⇒ hoist `badge-clickable` classname

rosamaggi committed:
[Fix] button style

rosamaggi committed:
[Fix] rename props

rosamaggi committed:
[Fix]Typo

rosamaggi committed:
[Feature] Add className prop

rosamaggi committed:
[FIX] Resolve dependencies with node module

rosamaggi committed:
[Fix] Add label and icon as props

rosamaggi committed:
[Refactor] Rename Card to FormCard

rosamaggi committed:
[Fix] Specialize style name with the epfl-elements-react

rosamaggi committed:
[Fix] Specify style name

rosamaggi committed:
[Feature] Add debounce input

rosamaggi committed:
[Install] p-debounce-throttle

rosamaggi committed:
[Fix] Make the debounce faster

rosamaggi committed:
[Fix] style

rosamaggi committed:
[Feature] Add style props

rosamaggi committed:
[Fix] tabs style

rosamaggi committed:
[Fix] Add mandatory key props

rosamaggi committed:
[Fix] Add id and key

rosamaggi commented:
Hello,
I added some correction to the branch. Could you verify please?
Thanks
Rosa

rosamaggi committed:
[Style] Add style and className props

rosamaggi committed:
[Fix] replace Form.Control (react-bootstrap) with normal html input tag

rosamaggi committed:
[Fix] Delete header menu

rosamaggi committed:
[Feature] Make menu responsive

rosamaggi committed:
[Fix] Click outside the menu to close it

rosamaggi committed:
[Fix] bug with responsive menu

rosamaggi committed:
[Fix] Add key to TextArea

rosamaggi committed:
[Refactor] eliminate dead code

rosamaggi committed:
[Feature] Add tooltip on button

rosamaggi committed:
[Fix] style

rosamaggi committed:
[Fix] Add autofocus
